README.md
# Flutter Global Animations
[![pub package](https://img.shields.io/pub/v/flutter_global_animations.svg)](https://pub.dev/packages/flutter_global_animations)

A flutter plugin to play custom global animations in your application.

## Example

![Example](https://raw.githubusercontent.com/jonasroussel/flutter_global_animations/master/example/example.gif)

## Usage

### Import
```dart
import 'package:flutter_global_animations/flutter_global_animations.dart';
```

### Create your own animation

```dart
class IconReaction extends AnimationModel {
  IconReaction({
    @required Offset position,
    @required Size region,
    @required IconData icon,
  }) : super(
          position: position,
          region: region,
          duration: Duration(milliseconds: 600),
          count: 10,
          data: icon,
        );

  final maxHeight = 300;
  final minDY = 0.3;
  final maxDX = 0.2;
  final minSize = 30;
  final maxSize = 80;

  CurvedAnimation _fadeCurve;
  CurvedAnimation _posCurve;

  @override
  AnimationController createController(TickerProvider vsync) {
    AnimationController controller = super.createController(vsync);

    _fadeCurve = CurvedAnimation(parent: controller, curve: Cubic(0.9, 0.0, 1.0, 1.0));
    _posCurve = CurvedAnimation(parent: controller, curve: Curves.easeOut);

    return controller;
  }

  @override
  Entity createEntity() {
    final entity = super.createEntity();

    double dy = -(random.nextDouble() * (1 - minDY) + minDY);

    entity.dy = dy;

    return entity;
  }

  @override
  Widget buildEntity(Entity entity) {
    return Opacity(
      opacity: 1 - _fadeCurve.value,
      child: Container(
        child: Directionality(
          textDirection: TextDirection.ltr,
          child: Icon(
            data,
            size: 24,
            color: Colors.lightBlue,
          ),
        ),
      ),
    );
  }

  @override
  Offset getEntityPosition(Entity entity) {
    double posX = entity.dx * (_posCurve.value * region.width);
    double posY = entity.dy * (_posCurve.value * maxHeight);

    return Offset(posX, posY);
  }
}
```

### Wrap your app in `GlobalAnimationsWrapper`

```dart
class ExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GlobalAnimationsWrapper( // <-- HERE
      child: MaterialApp(
        title: 'Example App',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          splashFactory: InkRipple.splashFactory,
        ),
        home: HomePage(),
      ),
    );
  }
}
```

### Spawn an animation

```dart
...
child: InkWell(
  onTap: () {
    RenderBox renderBox = context.findRenderObject();
    Size size = renderBox.size;
    Offset position = renderBox.localToGlobal(Offset.zero);

    GlobalAnimations.of(context).spawn(
      IconReaction(
        position: position,
        region: size,
        icon: icons[index],
      ),
    );
  },
...
```


CHANGELOG.md
## 1.0.0

- Stable release for null safety

## 0.1.2

- Example gif now github url

## 0.1.1

- Example gif in README.md

## 0.1.0

- First clean version, just what I supposed to be


lib/src/engine.dart
import 'dart:math';

import 'package:flutter/widgets.dart';

import 'models.dart';

class Animations extends ChangeNotifier {
  Animations(this._vsync);

  final TickerProvider _vsync;

  final List<AnimationModel> _animations = [];

  void spawn(AnimationModel animation) {
    final controller = animation.createController(_vsync);

    controller.addStatusListener((status) {
      if (status == AnimationStatus.completed) {
        _animations.remove(animation..dispose());
      }
    });
    controller.addListener(() {
      notifyListeners();
    });

    _animations.add(animation..start());
  }

  List<Widget> render() => _animations
      .map((anim) => anim.buildEntities())
      .expand((pair) => pair)
      .toList();

  @override
  void dispose() {
    _animations.forEach((anim) {
      anim.dispose();
    });
    super.dispose();
  }
}

abstract class AnimationModel {
  AnimationModel({
    required this.position,
    required this.region,
    required this.duration,
    this.count = 1,
    this.data,
  });

  final Offset position;
  final Size region;
  final Duration duration;
  final int count;
  final dynamic data;

  List<Entity>? entities;
  AnimationController? controller;

  AnimationController createController(TickerProvider vsync) {
    if (controller != null) return controller!;
    return (controller = AnimationController(vsync: vsync, duration: duration));
  }

  Entity createEntity() {
    final random = Random();

    final x = random.nextDouble() * region.width;
    final y = random.nextDouble() * region.height;

    return Entity(x: x, y: y);
  }

  void start() {
    entities = List.generate(count, (index) => createEntity());
    controller?.forward();
  }

  Iterable<Widget> buildEntities() {
    return (entities ?? []).map((entity) {
      final pos = getEntityPosition(entity);

      return Positioned(
        left: entity.x + position.dx + pos.dx,
        top: entity.y + position.dy + pos.dy,
        child: IgnorePointer(child: buildEntity(entity)),
      );
    });
  }

  void dispose() {
    controller?.dispose();
    entities?.clear();
  }

  Widget buildEntity(Entity entity);
  Offset getEntityPosition(Entity entity);
}


lib/src/wrapper.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

import 'engine.dart';

class GlobalAnimations extends InheritedWidget {
  static GlobalAnimations of(BuildContext context) =>
      context.dependOnInheritedWidgetOfExactType<GlobalAnimations>()!;

  GlobalAnimations(
    this._state, {
    Key? key,
    required Widget child,
  }) : super(key: key, child: child);

  final _GlobalAnimationsWrapperState _state;

  @override
  bool updateShouldNotify(InheritedWidget oldWidget) => false;

  void spawn(AnimationModel animation) => _state.animations.spawn(animation);
}

class GlobalAnimationsWrapper extends StatefulWidget {
  GlobalAnimationsWrapper({required this.child});

  final Widget child;

  @override
  _GlobalAnimationsWrapperState createState() =>
      _GlobalAnimationsWrapperState();
}

class _GlobalAnimationsWrapperState extends State<GlobalAnimationsWrapper>
    with TickerProviderStateMixin {
  late Animations animations;

  @override
  void initState() {
    super.initState();
    animations = Animations(this);
  }

  @override
  void dispose() {
    animations.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GlobalAnimations(
      this,
      child: AnimatedBuilder(
        animation: animations,
        builder: (context, child) {
          return Stack(
            alignment: Alignment.topLeft,
            children: [
              Positioned.fill(child: child!),
              ...animations.render(),
            ],
          );
        },
        child: widget.child,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_global_animations/flutter_global_animations.dart';

void main() {
  runApp(ExampleApp());
}

const icons = [
  Icons.ac_unit,
  Icons.alarm,
  Icons.account_balance,
  Icons.add_shopping_cart,
  Icons.airplanemode_active,
  Icons.alternate_email,
  Icons.album,
  Icons.add_circle_outline,
  Icons.outlined_flag,
  Icons.brightness_2,
  Icons.wifi,
  Icons.brightness_6,
];

class ExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GlobalAnimationsWrapper(
      child: MaterialApp(
        title: 'Example App',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          splashFactory: InkRipple.splashFactory,
        ),
        home: HomePage(),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Example App'),
      ),
      body: ListView.separated(
        padding: EdgeInsets.symmetric(vertical: 20),
        itemCount: icons.length,
        itemBuilder: (context, index) => Center(child: ItemCard(index: index)),
        separatorBuilder: (context, index) => SizedBox(height: 10),
      ),
    );
  }
}

class ItemCard extends StatelessWidget {
  ItemCard({this.index});

  final int index;

  @override
  Widget build(BuildContext context) {
    return Card(
      child: InkWell(
        onTap: () {
          RenderBox renderBox = context.findRenderObject();
          Size size = renderBox.size;
          Offset position = renderBox.localToGlobal(Offset.zero);

          GlobalAnimations.of(context).spawn(
            IconReaction(
              position: position,
              region: size,
              icon: icons[index],
            ),
          );
        },
        child: Padding(
          padding: EdgeInsets.symmetric(vertical: 8, horizontal: 16),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text('Button  '),
              Icon(icons[index]),
            ],
          ),
        ),
      ),
    );
  }
}

class IconReaction extends AnimationModel {
  IconReaction({
    @required Offset position,
    @required Size region,
    @required IconData icon,
  }) : super(
          position: position,
          region: region,
          duration: Duration(milliseconds: 600),
          count: 10,
          data: icon,
        );

  final maxHeight = 300;
  final minDY = 0.3;
  final maxDX = 0.2;
  final minSize = 30;
  final maxSize = 80;

  CurvedAnimation _fadeCurve;
  CurvedAnimation _posCurve;

  @override
  AnimationController createController(TickerProvider vsync) {
    AnimationController controller = super.createController(vsync);

    _fadeCurve =
        CurvedAnimation(parent: controller, curve: Cubic(0.9, 0.0, 1.0, 1.0));
    _posCurve = CurvedAnimation(parent: controller, curve: Curves.easeOut);

    return controller;
  }

  @override
  Entity createEntity() {
    final entity = super.createEntity();

    double dy = -(random.nextDouble() * (1 - minDY) + minDY);

    entity.dy = dy;

    return entity;
  }

  @override
  Widget buildEntity(Entity entity) {
    return Opacity(
      opacity: 1 - _fadeCurve.value,
      child: Container(
        child: Directionality(
          textDirection: TextDirection.ltr,
          child: Icon(
            data,
            size: 24,
            color: Colors.lightBlue,
          ),
        ),
      ),
    );
  }

  @override
  Offset getEntityPosition(Entity entity) {
    double posX = entity.dx * (_posCurve.value * region.width);
    double posY = entity.dy * (_posCurve.value * maxHeight);

    return Offset(posX, posY);
  }
}
